src/main/java/org/candlepin/controller/ContentAccessManager.java

```diff
@@ -804,4 +820,192 @@ public Owner syncOwnerLastContentUpdate(Owner owner) {
         return this.ownerCurator.merge(owner);
     }
 
+    /**
+     * Retrieves an existing {@link AnonymousContentAccessCertificate} for the {@link AnonymousCloudConsumer}
+     * or creates a new certificate if the existing certificate is expired or does not exist.
+     *
+     * @param consumer
+     *  the anonymous consumer to retrieve or create an {@link AnonymousContentAccessCertificate} for
+     *
+     * @throws IllegalArgumentException
+     *  if the provided anonymous cloud consumer is null
+     *
+     * @throws Exception
+     *  if unable to create a new {@link AnonymousContentAccessCertificate}
+     *
+     * @return
+     *  the retieved or generated certificate
+     */
+    @Transactional
+    public AnonymousContentAccessCertificate getCertificate(AnonymousCloudConsumer consumer)
+        throws Exception {
+        if (consumer == null) {
+            throw new IllegalArgumentException("anonymous cloud consumer is null");
+        }
+
+        AnonymousContentAccessCertificate cert = consumer.getContentAccessCert();
+        Date now = new Date();
+
+        // Generate a new certificate if one does not exist or the existing certificate is expired
+        if (cert == null || cert.getSerial() == null || cert.getSerial().getExpiration().before(now)) {
+            cert = createAnonContentAccessCertificate(consumer);
+        }
+
+        AnonymousContentAccessCertificate wrappedCert = new AnonymousContentAccessCertificate();
+        wrappedCert.setCert(cert.getCert() + cert.getContent());
+        wrappedCert.setCreated(cert.getCreated());
+        wrappedCert.setUpdated(cert.getUpdated());
+        wrappedCert.setId(cert.getId());
+        wrappedCert.setAnonymousCloudConsumer(cert.getAnonymousCloudConsumer());
+        wrappedCert.setKey(cert.getKey());
+        wrappedCert.setSerial(cert.getSerial());
+
+        return wrappedCert;
+    }
+
+    private AnonymousContentAccessCertificate createAnonContentAccessCertificate(
+        AnonymousCloudConsumer consumer) throws Exception {
+        log.info("Generating content access certificate for anonymous cloud consumer: {}",
+            consumer.getUuid());
+
+        List<ProductInfo> products = prodAdapter.getChildrenByProductIds(consumer.getProductIds());
+        if (products == null || products.isEmpty()) {
+            String msg = "Unable to retrieve products for anonymous cloud consumer: " + consumer.getUuid();
+            throw new Exception(msg);
```

Comment:
Please, never do this. Throwing `Exception` directly breaks any usefulness the Java's exceptions have.

- It forces all callers to declare `throws Exception` which hides all information about what the method actually throws.
- If the the exception needs to be handled it forces the caller into pokemon exception handling. Causing it to potentially catch unwanted exceptions (such as NPE).

Also it should probably be unchecked exception. It could be checked exception, but it won't be handled. Just propagated to the endpoint and translated to http error, so it would only pollute method definitions along the way. Either way it should be our exception to signify certificate generation failed.

Reply:
Yeah, those are great points! Updated